allensdk/brain_observatory/dff.py

```diff
@@ -232,7 +242,129 @@ def compute_dff(traces, save_plot_dir=None, mode_kernelsize=5400, mean_kernelsiz
 
         logging.debug("finished trace %d/%d" % (n + 1, traces.shape[0]))
 
-        if save_plot_dir:
+    return dff
+
+
+def compute_dff_windowed_median(traces,
+                                median_kernel_long=5401,
+                                median_kernel_short=101,
+                                noise_kernel_length=31):
+    """Compute dF/F of a set of traces with median filter detrending.
+
+    The operation is basically:
+
+        T_long = windowed_median(T) # long timescale kernel
+
+        T_dff1 = (T - T_long) / elementwise_max(T_long, noise_std(T))
+
+        T_short = windowed_median(T_dff1) # short timescale kernel
+
+        T_dff = T_dff1 - elementwise_min(T_short, 2.5*noise_std(T_dff1))
+
+    Parameters
+    ----------
+    traces : np.ndarray
+       2D array of traces to be analyzed.
+    median_kernel_long : int
+        Window size to use for long timescale median detrending.
+    median_kernel_short : int
+        Window size to use for short timescale median detrending.
+    noise_kernel_length : int
+        Window size to use for high pass filtering noise.
+
+    Returns
+    -------
+    dff : np.ndarray
+        2D array of dF/F traces.
+    """
+    for k in [median_kernel_long, median_kernel_short, noise_kernel_length]:
+        _check_kernel(k, traces.shape[1])
+
+    dff_traces = np.copy(traces)
+
+    for dff in dff_traces:
+        sigma_f = noise_std(dff, window_length=noise_kernel_length)
+
+        # long timescale median filter for baseline subtraction
+        tf = median_filter(dff, median_kernel_long, mode='constant')
+        dff -= tf
+        dff /= np.maximum(tf, sigma_f)
+
+        sigma_dff = noise_std(dff, window_length=noise_kernel_length)
+
+        # short timescale detrending
+        tf = median_filter(dff, median_kernel_short, mode='constant')
+        tf = np.minimum(tf, 2.5*sigma_dff)
+        dff -= tf
+
+    return dff_traces
+
+
+def _check_kernel(kernel_size, data_size):
+    if kernel_size % 2 == 0 or kernel_size <= 0 or kernel_size >= data_size:
+        raise ValueError("Invalid kernel length {} for data length {}. Kernel "
+                         "length must be positive and odd, and less than data "
+                         "length.".format(kernel_size, data_size))
+
+
+def noise_std(x, window_length=31):
+    """Robust estimate of the standard deviation of the trace noise."""
+    if any(np.isnan(x)):
+        return np.NaN
+    x = x - median_filter(x, window_length, mode='constant')
+    # first pass removing big pos peak outliers
+    x = x[x < 1.5*np.abs(x.min())]
```

Comment:
Should you expose these two params as kwargs?

Reply:
Yes, I think those should be exposed. I have a few more questions for the analysis developers before I finalize this as well.